grpcutil/dns_resolver.go

```diff
@@ -131,7 +131,7 @@ func (r *dnsResolver) Resolve(target string) (Watcher, error) {
 
 // dnsWatcher watches for the name resolution update for a specific target
 type dnsWatcher struct {
```

Comment:
I would remove `Watcher` too and expose `dnsWatcher` as Watcher. WDYT? Nothing in dskit needs `Watcher` interface.

Reply:
https://github.com/grafana/dskit/pull/102#discussion_r774620029

From @aknuds1:
> I realized that the reason for the Watcher interface is that there are two different implementations: dnsWatcher and ipWatcher. I.e., getting rid of the interface is trickier. Do you still think we should do it?

Reply:
@pstibrany The problem is we have a second implementation of `Watcher` that can also be returned from `Resolve`: `ipWatcher`.

Reply:
You're right, `Watcher` hides the implementation. I should have checked it better. Let's keep it.

Reply:
Implementation wise, we can solve it without an interface (fusing the two structs), but do we want to?

Reply:
I don't think we want to merge them.